File: Problem117_h-index.java
// Time Complexity: O(n)
// Space Complexity: O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


class Solution {
    public int hIndex(int[] citations) {
        
        int n = citations.length;
        int count[] = new int[n + 1];
        for(int i = 0; i < n; i++) {
            if(citations[i] >= n) {
                count[n]++;
            } else {
                count[citations[i]]++;
            }
        }
        int sum = 0;
        for(int i = count.length - 1; i >= 0; i--) {
            sum += count[i];
            if(i <= sum)
                return i;
        }
        return 0;
    }
}

File: Problem-117 H-Index.py
# 274. H-Index
# https://leetcode.com/problems/h-index/

# Logic: Initialize a count array. Iterate over the input and 
# for every element update the count. If the element value is 
# more than n, update the element at nth position. Iterate the 
# count from the end and check if the count is greater than 
# equals index, return the index.

# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution:
    def hIndex(self, citations) -> int:
        
        n = len(citations)
        
        count = [0 for i in range(n+1)]
        
        for i in citations:
            if i >= n:
                count[n] += 1
            else:
                count[i] += 1
        
        papers = 0
        for i in range(n, -1, -1):
            papers += count[i]
            if papers >= i:
                return i